### PR TITLE
Exclude planner & plan evaluator from tests for features not yet implemented

### PR DESCRIPTION
**Pull request by dlurton**
Originally part of #582  and based on #583 (which must be merged first), this PR primarily sets the target evaluator for all tests related to the following features to be the AST evaluator, thereby excluding the planner and plan evaluator from them.  (These features are not within the current scope of work for the planner and new evaluator.)  Note that the new planner and evaluator are not in this branch yet and will be introduced in a separate pull request.  There are also a few new tests that I added to provide simpler test cases for a few scenarios that were a little difficult to get working with the new planner and evaluator.

- `ORDER BY`
- `GROUP BY`
- `HAVING`
- Aggregate functions (`SUM`, `MAX`, `MIN`, etc)
- `PIVOT`
- `UNPIVOT`

Also, some of the exception tests have been set to only execute under the `CompilerPipeline` because there is a behavior change surrounding a rare edge case: "ambiguous bindings", i.e. when you have a case-insensitive variable reference and more than one matching variable that varies only by letter case.  Example: if you have `fOo` and `FoO` defined but you refer to variable `foo` (case-insenstive), `CompilerPipeline` etc would throw an exception at evaluation time if both are defined in the same lexical scope.  The plan evaluator simply selects `fOo` and moves along merrily.  (This behavior is consistent with Postgres, BTW.)   This change is fully backward compatible because it allows queries to execute that wouldn't previously.

The `PlannerPipeline` does not support `UndefinedVariableBehavior.MISSING`, so tests related that also execute only under `CompilerPipeline`.  This feature was intended for backward compatibility for a legacy customer who is stuck on a very old version of PartiQL for other reasons anyway.

Lastly, the `PlannerPipeline` does not (yet?) support the `StaticTypeInferenceVisitorTransform`, so any test that required the inferencer (i.e. `EvaluatingCompilerNAryIntOverflowTests`) to run has also been set to only run against the `CompilerPipeline`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt [hunk 1]
```diff
@@ -439,6 +439,52 @@ internal val EVALUATOR_TEST_SUITE: IonResultTestSuite = defineTestSuite {
...
+            """,
+            """
+              $partiql_bag::[
+                {id: "5"},
+                {id: "6", title: "E"},
+                {id: "6", title: "F"},
+                {id: "7"}
+              ]
+            """
+        )
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt:486]: These tests seem repeated:
- `selectJoinOnClauseScoping` -- [L488-506](https://github.com/partiql/partiql-lang-kotlin/blob/ea26eea106d5a5b7dd689a473daf1539347d0c69/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L488-L506)
- `selectCorrelatedLeftJoin` -- [L415-423](https://github.com/partiql/partiql-lang-kotlin/blob/ea26eea106d5a5b7dd689a473daf1539347d0c69/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L415-L423)
- `selectCorrelatedLeftJoinOnClause` -- [L425-440](https://github.com/partiql/partiql-lang-kotlin/blob/ea26eea106d5a5b7dd689a473daf1539347d0c69/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L425-L440)

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 2]
```diff
@@ -21,6 +21,7 @@ import org.junit.jupiter.params.provider.ArgumentsSource
 import org.partiql.lang.errors.ErrorCode
 import org.partiql.lang.errors.Property
 import org.partiql.lang.eval.evaluatortestframework.EvaluatorErrorTestCase
+import org.partiql.lang.eval.evaluatortestframework.EvaluatorTestTarget
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:24]: In order to fully replace `CompilerPipeline` and `EvaluatingCompiler`--yes.  But when and who will implement them has not been decided.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 3]
```diff
@@ -99,28 +99,32 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
         """SELECT VALUE a FROM `[{v:5}]` AS item, @item.v AS a, @item.v AS a""",
         ErrorCode.EVALUATOR_AMBIGUOUS_BINDING,
         expectedErrorContext = propertyValueMapOf(1, 14, Property.BINDING_NAME to "a", Property.BINDING_NAME_MATCHES to "a, a"),
-        expectedPermissiveModeResult = "<<MISSING>>"
+        expectedPermissiveModeResult = "<<MISSING>>",
+        target = EvaluatorTestTarget.COMPILER_PIPELINE
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:103]: You're correct, sir. Good catch.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 4]
```diff
@@ -140,7 +144,8 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
         // Same query as previous test--but DO NOT throw exception this time because of UndefinedVariableBehavior.MISSING
         runEvaluatorTestCase(
             sqlWithUndefinedVariable, expectedResult = "[null]",
-            compileOptionsBuilderBlock = { undefinedVariable(UndefinedVariableBehavior.MISSING) }
+            compileOptionsBuilderBlock = { undefinedVariable(UndefinedVariableBehavior.MISSING) },
+            target = EvaluatorTestTarget.COMPILER_PIPELINE
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:148]: yes, probably got a little overzealous with my copy and pasting.  Thanks.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:103]: Updated PR description to include this.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 5]
```diff
@@ -160,7 +165,8 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
         // Same query as previous test--but DO NOT throw exception this time because of UndefinedVariableBehavior.MISSING
         runEvaluatorTestCase(
             sqlWithUndefinedQuotedVariable, expectedResult = "[null]",
-            compileOptionsBuilderBlock = { undefinedVariable(UndefinedVariableBehavior.MISSING) }
+            compileOptionsBuilderBlock = { undefinedVariable(UndefinedVariableBehavior.MISSING) },
+            target = EvaluatorTestTarget.COMPILER_PIPELINE
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:169]: answered above.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt [hunk 6]
```diff
@@ -187,14 +193,16 @@ class EvaluatingCompilerExceptionsTest : EvaluatorTestBase() {
...
+        target = EvaluatorTestTarget.COMPILER_PIPELINE
     )
 
     @Test
     fun outerJoin() = runEvaluatorErrorTestCase(
         "SELECT * FROM animals AS a OUTER CROSS JOIN animal_types AS a_type WHERE a.type = a_type.id",
         ErrorCode.EVALUATOR_FEATURE_NOT_SUPPORTED_YET,
-        expectedErrorContext = propertyValueMapOf(1, 28, Property.FEATURE_NAME to "RIGHT and FULL JOIN")
+        expectedErrorContext = propertyValueMapOf(1, 28, Property.FEATURE_NAME to "RIGHT and FULL JOIN"),
+        target = EvaluatorTestTarget.COMPILER_PIPELINE
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:205]: I assume you mean:

> Will features not currently supported in the `COMPILER_PIPELINE` also set the target to **`PLANNER_PIPELINE`** then?

And if so, then yes.  However, IIRC there not any yet.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerNAryIntOverflowTests.kt [hunk 7]
```diff
@@ -164,7 +162,8 @@ class EvaluatingCompilerNAryIntOverflowTests : EvaluatorTestBase() {
                 compileOptions {
                     typingMode(TypingMode.PERMISSIVE)
                 }
-            }
+            },
+            target = EvaluatorTestTarget.COMPILER_PIPELINE // Requires PlannerPipeline to support globalTypeBindings and static type inference
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerNAryIntOverflowTests.kt:166]: These tests only work when the static type inferencer is enabled which is not yet supported by `PlannerPipeline`.  Updated PR description and more explicitly stated this in the comment.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerOffsetTests.kt [hunk 8]
```diff
@@ -105,7 +106,10 @@ class EvaluatingCompilerOffsetTests : EvaluatorTestBase() {
     @ParameterizedTest
     @ArgumentsSource(ArgsProviderValid::class)
     fun validTests(tc: EvaluatorTestCase) = runEvaluatorTestCase(
-        tc.copy(excludeLegacySerializerAssertions = true),
+        tc.copy(
+            excludeLegacySerializerAssertions = true,
+            target = EvaluatorTestTarget.COMPILER_PIPELINE, // planner & phys. alg. have no support for OFFSET (yet)
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerOffsetTests.kt:111]: This comment was wrong actually (`OFFSET` is supported by `PlannerPipeline`), but some of the test cases inside used `GROUP BY` and had to be set to `COMPILER_PIPELINE`.  Thanks.

File: lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt (hunk 1 above)
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt:486]: Oops, thanks. Removing...

File: lang/test/org/partiql/lang/eval/QuotedIdentifierTests.kt [hunk 9]
```diff
@@ -17,6 +17,7 @@ package org.partiql.lang.eval
 import org.junit.Test
 import org.partiql.lang.errors.ErrorCode
 import org.partiql.lang.errors.Property
+import org.partiql.lang.eval.evaluatortestframework.EvaluatorTestTarget
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/QuotedIdentifierTests.kt:20]: previously done.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:103]: nit: may be worth leaving a comment for this test stating what you mentioned in this comment (or link the PR comment/description)

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerExceptionsTest.kt:148]: Thanks for editing the PR description for this.

nit: could also leave a comment for this test and the one on L169 stating what's mentioned here. Or could use a comment you had in `QuotedIdentifierTests.kt`

```
// planner & physical plan have no support for UndefinedVariableBehavior.MISSING (and may never)
```

**Review by alancai98 — APPROVED** [review]
Thanks for providing more details in the PR description. Changes look good. Only have a couple minor nits to include some of the PR conversation in a comment.
